Implement the fallback entity lookup

dsenalik opened the pull request:
# New Feature

### Closes #1850 

### Tripal Version: 4.x

## Description
There are four cases in core Tripal where links use a different name, currently entity links do not work for these.
The goofy fields that we need to test are listed at the end of this comment https://github.com/tripal/tripal_doc/issues/32#issuecomment-1837532399
```
arraydesign FOREIGN KEY (manufacturer_id) REFERENCES contact(contact_id)
assay FOREIGN KEY (operator_id) REFERENCES contact(contact_id)
biomaterial FOREIGN KEY (biosourceprovider_id) REFERENCES contact(contact_id)
biomaterial FOREIGN KEY (taxon_id) REFERENCES organism(organism_id)
```

We had originally intended to consult the third party settings, but for reasons of efficiency I decided to have the TripalTypes property pass through the table name. It already knows this, and thus avoids a chunk of code to get at the third party settings to look it up.

Because I changed the entity_id property, we need an update function. Again for efficiency, I recycled the existing update function, so as not to duplicate it, it will just be run again with a higher number.

## Testing?
 1. Create your docker on the 4.x branch so we can test the update function
 2. Change to this branch `git checkout tv4g1-issue1850-entity-lookup-third-party-settings`
 3. Run the update `docker updatedb` - verify no errors are reported
 4. Import the gene expression type collection
 5. Create an organism
 6. Create a contact
 7. Create an array design. You will need a contact that you created in step 6
 8. Verify that the "Manufacturer" is a link 
 9. Create an assay. Use the contact as the "Operator"
 10. Again verify that the "Operator" is a link
 11. Create a biomaterial.
 12. Verify that both biosource provider and Organism (taxon id) are links
 13. I have added a kernel test for the fallback method (manufacturer on arraydesign) so make sure tests pass 🙌

dsenalik committed:
add description header to TripalEntityLookup class

dsenalik committed:
fix so far just for contact field

dsenalik committed:
revise the update function

dsenalik committed:
Update remaining entity_id fields

dsenalik committed:
make a bit more efficient

dsenalik committed:
add tests for fallback entity lookup

dsenalik committed:
Merge branch '4.x' into tv4g1-issue1850-entity-lookup-third-party-settings

dsenalik committed:
Merge branch '4.x' into tv4g1-issue1850-entity-lookup-third-party-settings

laceysanderson reviewed (commented):
Sorry for the delay on this 🙈 

I have a couple of comments/questions:
1. I like that you are setting the foreign key table in the field settings now! That is a great approach and means we need to use the fallback checking all entities less often. ❤️ 
2. For the fallback, I see you are only grabbing a single backup entity type, whereas we had discussed looping through all content types where that is a base table. I still think what we discussed is a better approach 🤔 but am wondering why you decided to just focus on one?
3. I see that you deleted a migration 😬 Can you explain why you did that? I'm worried about people who did apply that original migration 🤔  although I doubt there are that many...

dsenalik committed:
explain better why update 10403 was removed

dsenalik commented:
For item 2: Thanks @laceysanderson for calling me out on that function, what it really did was just confirm that a bundle existed with the same name as the base table, and as you noted that was not what we planned so was not correct.

Here is my debugging code for the correct way to retrieve all the bundles with a given base table
```
// Should retrieve bundle = 'array_design' given base table 'arraydesign'
$base_table = 'arraydesign';

$bundles = \Drupal::entityTypeManager()
  ->getStorage('tripal_entity_type')
  ->getQuery()
  ->condition('third_party_settings.tripal.chado_base_table', $base_table)
  ->execute();
var_dump($bundles);
```
and this returns
```
array(1) {
  ["array_design"]=>
  string(12) "array_design"
}
```
Now for the `feature` table for example, no bundle name will match the base table name, so I will be implementing this as follows:
1. If nothing is returned, return null

2. If exactly one bundle is returned, return that bundle (for example base table = "arraydesign") even if bundle name ("array_design") does not match base table

  3. If more than one bundle is returned (for example base table = "analysis") {
  4. If one of the bundle names matches the base table name, then return that
  } else {
  5. Return null, because there is no "Default" bundle (for example base table="feature")
  There is no current Chado table which will cause us to get to this case, fortunately,
  }

dsenalik committed:
redesign getDefaultBundle()

dsenalik commented:
For Item # 3, actually I did not remove the migration, I effectively renumbered it, with some additions. That is not very obvious looking at the github changes, but here is the diff. Since the properties may have been added with 10403, or may be added for the first time with 10405, it now handles both cases.
But the short version is that if you never ran 10403 then 10405 will do everything needed.
I did add some better comments to explain this with commit https://github.com/tripal/tripal/pull/1865/commits/17fdcc657d75373d99eb9b7aace98a378e6575ec
```
4c5
< function tripal_chado_update_10403() {
---
> function tripal_chado_update_10405() {
32a34
>     $n_updated = 0;
48c50,56
<               if (!$field_exists) {
---
>               if ($field_exists) {
>                 $messenger->addMessage(t("Updating entity id in Drupal table \"@table_name\", field \"@field_name\"",
>                     ['@table_name' => $table_name, '@field_name' => $field_name]));
>                 $schema->changeField($table_name, $columns[$new_property], $columns[$new_property], $field_schema['columns'][$new_property]);
>                 $n_updated++;
>               }
>               else {
52d59
<                 $manager->updateFieldStorageDefinition($field_storage_definition);
57a65
>         $manager->updateFieldStorageDefinition($field_storage_definition);
60c68,69
<     $messenger->addMessage(t("Added @n_added entity id properties", ['@n_added' => $n_added]));
---
>     $messenger->addMessage(t("Added @n_added entity id properties, updated @n_updated entity id properties",
>         ['@n_added' => $n_added, '@n_updated' => $n_updated]));
```

dsenalik committed:
update test and add third party setting to schema

dsenalik commented:
Commit https://github.com/tripal/tripal/pull/1865/commits/2004d64d46ca2a2f2144dc0b0846e2bdef689951 fixes the following unit test error
`Drupal\Core\Config\Schema\SchemaIncompleteException: Schema errors for tripal.content_type.contact with the following errors: tripal.content_type.contact:third_party_settings.tripal.chado_base_table missing schema`
and includes a new test of `setThirdPartySetting()` + `getThirdPartySetting()`
This is because with the correct implementation of `getDefaultBundle()` we need to have the third party setting for `chado_base_table` set, but it was not set by default in the testing environment, so that had to be added.

laceysanderson commented:
Regarding 
> If more than one bundle is returned (for example base table = "analysis") {
>    a) If one of the bundle names matches the base table name, then return that
>    b) Return null, because there is no "Default" bundle (for example base table="feature") There is no current Chado table which will cause us to get to this case, fortunately.

I'm confused by what you mean when you say no current chado table will get to the case where there is no "Default" bundle. Wouldn't we hit this case for any field on a feature-based content type that does not have the foreign key table set?

As an alternative, maybe you should rename the function to indicate that it will return all possible bundles. Then in getEntityId() you loop through the array of possible bundles, checking each one until you find it?

laceysanderson commented:
> Commit [2004d64](https://github.com/tripal/tripal/commit/2004d64d46ca2a2f2144dc0b0846e2bdef689951) fixes the following unit test error `Drupal\Core\Config\Schema\SchemaIncompleteException: Schema errors for tripal.content_type.contact with the following errors: tripal.content_type.contact:third_party_settings.tripal.chado_base_table missing schema` and includes a new test of `setThirdPartySetting()` + `getThirdPartySetting()` This is because with the correct implementation of `getDefaultBundle()` we need to have the third party setting for `chado_base_table` set, but it was not set by default in the testing environment, so that had to be added.

For this addition, maybe we should move this into the test trait createContentTypeFromConfig() function so that all tests have access to the new third party setting and better represent the live environment?

dsenalik committed:
fallback checks each possible bundle; remove unused function

dsenalik committed:
move third party setting to TripalTestTrait

dsenalik commented:
Thanks for keeping me honest @laceysanderson 
  > As an alternative, maybe you should rename the function to indicate that it will return all possible bundles. Then in getEntityId() you loop through the array of possible bundles, checking each one until you find it?

Implemented this with commit https://github.com/tripal/tripal/pull/1865/commits/8aa263af9599d87ec25157b425757a13910a79c5

  > maybe we should move this into the test trait createContentTypeFromConfig() function so that all tests have access to the new third party setting and better represent the live environment?

Another great suggestion, implemented this with commit https://github.com/tripal/tripal/pull/1865/commits/dd7012ac531040a57a2308494e59365b76fb5c17

laceysanderson reviewed (approved):
✅ Code changes look perfect :-)
✅ Manual testing worked great!

![image](https://github.com/tripal/tripal/assets/1566301/e7f1f7ed-724c-411a-84a1-9443365b5ec4)
![image](https://github.com/tripal/tripal/assets/1566301/6c26e11f-83aa-4bac-8f5a-409277611245)
![image](https://github.com/tripal/tripal/assets/1566301/cddc7653-0ce3-46bc-b562-9e91b6fa30dd)
![image](https://github.com/tripal/tripal/assets/1566301/93a90257-77ca-4551-98b7-0afdb863b385)

Thanks for the work @dsenalik!!! This is ready to merge! 😍 🥕
